### DISCUSSION TITLE
PIO Pull up, Pull down, none

### DISCUSSION THREAD

**Discussion by LordStress** [micropython/micropython#11329] [RP2040 / Pico]
Firstly I hope I am in the right place as I came across from MicroPython Forum (Archive)] (http://forum.micropython.org/)

Anyhow, I am using the PIO on the pico to read a 6 bit rotary encoder.  After a bit of reading and experimenting I deduced that e the GPIO pins seem to default to no internal pullup resistors being enabled.  So I needed to enable them explicitly for each of the 6 pins. As in:
``` py
# PIO Input test for reading data from a 6 bit rotary encoder
from machine import Pin
from rp2 import PIO, StateMachine, asm_pio

#Set up the input pins with pull up resistors    
Pin (2,Pin.IN,Pin.PULL_UP)
Pin (3,Pin.IN,Pin.PULL_UP)
Pin (4,Pin.IN,Pin.PULL_UP)
Pin (5,Pin.IN,Pin.PULL_UP)
Pin (6,Pin.IN,Pin.PULL_UP)
Pin (7,Pin.IN,Pin.PULL_UP)

#PIO code to read 6 data bits and put them on the RX FIFO
@asm_pio()
def Encoder_Read():
    in_(pins, 6)
    push()
    
#Set up the state machine and turn it on
#EncoderSM = rp2.StateMachine(1, Encoder_Read, freq = 2000, in_base=Pin(2, Pin.IN))  Is wrong as the Pin.IN removed the PULL_UP
EncoderSM = rp2.StateMachine(1, Encoder_Read, freq = 2000, in_base=Pin(2))  # This now works
EncoderSM.active(1)

#read encoder data from PIO FIFO
Value = EncoderSM.get() 
print (Value)
```

My question: Is there a more concise way of setting multiple PULL_UP resistors for multiple pins, other than doing them one at a time like my test program.

I also found that whereas the Pin.PULL_UP and Pin.PULL_DOWN are implemented for the RP2040 in MicroPython, the Pin.PULL_NONE is missing, although an absence of the PULL_UP/DOWN argument seems to default the pin PULL_NONE state.

** I edited the post as  one of the lines of code had a bug in it.

**Comment by GitHubsSilverBullet** [comment]
How about:
```py
pin_list = [Pin(n, Pin.IN, Pin.PULL_UP) for n in range(2,7+1)]
```
That way you can refer to them later-on.

**Reply by LordStress** [reply]
Yes thanks for that, and it does work, I had something similar:
``` py
for i in range(6):
    Pin (2+i,Pin.IN,Pin.PULL_UP)
``` 
Note you don't need to assign it to a list (in both methods).

I was hoping for something more along the lines of:
``` py
@asm_pio(out_init=(rp2.PIO.OUT_LOW ,) * 9, out_shiftdir=PIO.SHIFT_RIGHT)
``` 
Which I am using for my parallel output state machine, the *6 sets 6 pins.  There doesn't seem to be the equivalent when using the PIO for inputs. I realise it's not quite the same thing, as that is in a initialisation statement, but I think you get my drift.

Thanks for you help

**Reply by GitHubsSilverBullet** [reply]
Well, ›out_init‹ accepts a tuple, while Pin doesn't. Also keep in mind that ›out_init‹ doesn't know about the GPIO numbers. That's managed in the ›StateMachine‹ initialization somewhere else. And it is only allowed for consecutive pins.
There's no gain of speed or readability if it were all mixed together. 
And less than a one-liner to achieve something is very rarely seen.  ;-)
